Userland/useradd.cpp

```diff
@@ -0,0 +1,114 @@
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "home_dir", "home directory for the new user");
+    args_parser.add_arg("u", "uid", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "group", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "shell", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "name", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return 0;
+    }
+
+    const char* username = argv[argc - 1]; // Always the user name!
+    uid_t uid = BASE_UID;
+    gid_t gid = USERS_GID;
+
+    // First, let's sort out the uid for the user
+    if (args.is_present("u")) {
+        bool ok;
+        uid = static_cast<uid_t>(args.get("u").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid uid %s\n", args.get("u").characters());
+            return 3;
+        }
+
+        if (getpwuid(uid)) {
+            fprintf(stderr, "uid %u already exists!\n", uid);
+            return 4;
+        }
+
+    } else {
+        uid = BASE_UID;
+        for (; getpwuid(uid); uid++) {
+        }
+    }
+
+    // Now let's sort out the gid!
+    if (args.is_present("g")) {
+        bool ok;
+        gid = static_cast<gid_t>(args.get("g").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid gid %s\n", args.get("g").characters());
+            return -1;
+        }
+    }
+
+    String shell = args.is_present("s") ? args.get("s") : String(DEFAULT_SHELL);
+    String home = args.is_present("d") ? args.get("d") : String("/");
+    String gecos = args.is_present("n") ? args.get("n") : String("Anonymous");
+
+    if (args.is_present("m")) {
+        StringBuilder builder;
+        builder.append("/home/");
+        builder.append(username);
+
+        home = builder.to_string(); // This overwrites the -d argument
+        if (mkdir(home.characters(), 0700)) {
+            perror("mkdir");
+            return 12;
+        }
+
+        if (chown(home.characters(), uid, gid)) {
+            perror("chown");
+
+            if (rmdir(home.characters())) {
+                perror("rmdir");
```

Comment:
Same as below, print what is it that you failed to `mkdir`/`chown`/`rmdir`.

You don't need to do that for, say, `putpwent()`, because there's not much to add, and because it's clear that `useradd` calls `putpwent()` once. But as we can see, it deals with several files and directories, so better specify which one.